edb/pgsql/deltafts.py

```diff
@@ -0,0 +1,471 @@
+    return ops
+
+
+def _delete_fts_document(
+    index: s_indexes.Index,
+    drop_index: dbops.Command,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert subject
+    table_name = common.get_backend_name(schema, subject, catenate=False)
+
+    ops = dbops.CommandGroup()
+    ops.add_command(drop_index)
+
+    from edb.common import debug
+
+    if debug.flags.zombodb:
+        zombo_func_name = _zombo_func_name(table_name)
+        ops.add_command(dbops.DropFunction(zombo_func_name, args=[table_name]))
+
+        zombo_type_name = _zombo_type_name(table_name)
+        ops.add_command(dbops.DropCompositeType(zombo_type_name))
+    else:
+        ops.add_command(_pg_drop_trigger(table_name))
+
+        fts_document = dbops.Column(
+            name=f'__fts_document__',
+            type=('pg_catalog', 'tsvector'),
+        )
+        alter_table = dbops.AlterTable(table_name)
+        alter_table.add_operation(dbops.AlterTableDropColumn(fts_document))
+        ops.add_command(alter_table)
+    return ops
+
+
+def _refresh_fts_document(
+    index: s_indexes.Index,
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_types.Type)
+    table_name = common.get_backend_name(schema, subject, catenate=False)
+
+    # compile the expression
+    index_sexpr: Optional[s_expr.Expression] = index.get_expr(schema)
+    assert index_sexpr
+    index_expr = index_sexpr.ensure_compiled(
+        schema=schema,
+        options=options,
+    )
+
+    subject_id = irast.PathId.from_type(schema, subject)
+    sql_res = compiler.compile_ir_to_sql_tree(
+        index_expr.irast,
+        singleton_mode=True,
+        external_rvars={
+            (subject_id, 'source'): pgast.RelRangeVar(
+                alias=pgast.Alias(aliasname='NEW'),
+                relation=pgast.Relation(name='NEW'),
+            )
+        },
+    )
+    exprs = astutils.maybe_unpack_row(sql_res.ast)
+
+    ops = dbops.CommandGroup()
+
+    from edb.common import debug
+    if debug.flags.zombodb:
+        raise NotImplementedError('zombo refresh index not implemented')
+    else:
+        ops.add_command(_pg_drop_trigger(table_name))
+        ops.add_command(_pg_create_trigger(table_name, exprs))
+
+    return ops
+
+
+def _raise_unsupported_language_error(
+    unsupported: Collection[str],
+) -> None:
+    unsupported = list(unsupported)
+    unsupported.sort()
+
+    msg = 'Full text search language'
+    if len(unsupported) > 1:
+        msg += 's'
+
+    msg += ' ' + ', '.join(f'`{l}`' for l in unsupported)
+    msg += ' not supported'
+
+    raise errors.UnsupportedFeatureError(msg)
+
+
+# --- pg fts ---
+
+
+def _pg_create_trigger(
+    table_name: Tuple[str, str],
+    exprs: Sequence[pgast.BaseExpr],
+) -> dbops.Command:
```

Comment:
This needs to also update `__fts_document__` for all the existing rows, right?

Reply:
(Add a test for that case, too)

Reply:
Oh, yea it does. But not `_pg_create_trigger`, `_refresh_fts_document` should.

Reply:
Wait, no you are right - it has to be updated when creating the column AND when refreshing.